### PR TITLE
Sync 3d panel camera states

### PR DESCRIPTION
**Pull request by foxymiles**
**User-Facing Changes**
Adds a new panel setting to sync camera states across multiple 3d panels.

**Description**
This adds a new setting and new panel extension context features to sync the camera states of multiple 3d panels. This is accomplished via a new `sharedPanelState` property in the `PanelExtensionContext` `RenderState`.

Panels that wish to use shared state can watch `sharedPanelState` in the render state and call `setSharedPanelState` to update the shared state, which is typed as an opaque record type.

```typescript
    setSharedPanelState: (state: undefined | Readonly<string, unknown>) => void;
```

The shared state is scoped to the panel type, so, for example, the 3d panel will only see shared state for the 3d panel.

Screenshot 2022-11-21 at 12 08 13 PM

https://user-images.githubusercontent.com/93935560/202955310-91eeb167-b6b5-4bd9-a21c-74d410e781d7.mov

Fixes https://github.com/foxglove/studio/issues/4846

### CODE REVIEW

File: packages/studio/index.d.ts [hunk 1]
```diff
@@ -136,6 +136,19 @@ declare module "@foxglove/studio" {
...
+    perspective: boolean;
+    phi: number;
+    target: readonly [number, number, number];
+    targetOffset: readonly [number, number, number];
+    targetOrientation: readonly [number, number, number, number];
+    thetaOffset: number;
+    fovy: number;
+    near: number;
+    far: number;
+  };
```
Comment (jtbandes) [inline packages/studio/index.d.ts:150]: I'm not a fan of this being part of the public extension API. This couples the 3D panel implementation to the extension API and prevents us from making changes in the 3D panel camera system later (for example, https://github.com/foxglove/studio/issues/2081). We should strive to make the extension API powerful enough that panels as complex as 3D could be built as third-party extensions.

**Comment by jtbandes** [comment]
> 1. Should we sync the entire camera state or just pan/zoom?
> 2. What should we do if we have a mix of perspective and orthographic views? Sync them separately?

Entire state is how it worked in the old panel so I think that's fine for now. Perspective/ortho mode is part of the camera state, so that should be synced as well.

Sam has a good point regarding follow modes. We should probably sync the follow mode too, or disallow syncing the camera state if a different follow mode/display frame is used, because the syncing wouldn't really make sense.

File: packages/studio/index.d.ts (hunk 1 above)
Comment (foxymiles) [reply packages/studio/index.d.ts:150]: Do you have any alternative suggestions for exposing this in a way that other panels can consume it? Mark it as unknown instead and the panels cast it to the known type?

**Comment by jtbandes** [comment]
From Slack discussion, we have consensus on exposing a key-value-like API, similar to globalVariables, but where the variables are internal-only and not displayed in the sidebar. This allows panels to communicate, allows newly appearing panels to read current state, and avoids exposing specific 3D panel types in the extension API.

File: packages/studio-base/src/providers/PanelStateContextProvider.tsx [hunk 2]
```diff
@@ -61,8 +73,43 @@ export function usePanelSettingsTreeUpdate(): (newTree: ImmutableSettingsTree) =
...
+
 export function PanelStateContextProvider({ children }: { children?: ReactNode }): JSX.Element {
   const [store] = useState(createPanelStateStore());
 
+  const currentLayoutId = useCurrentLayoutSelector(selectCurrentLayoutId);
+
+  // clear shared panel state on layout change
+  useEffect(() => {
+    void currentLayoutId;
+    store.setState({ sharedPanelState: {} });
```
Comment (snosenzo) [inline packages/studio-base/src/providers/PanelStateContextProvider.tsx:111]: would it also be possible to clear this if there are no longer panels using it? I'm a little worried about it keeping references to panels around if extension panels should choose to somehow store a reference to itself, either intentionally or unintentionally

**Comment by defunctzombie** [comment]
Watching the video I think it might be nice to add an icon into the viewport to indicate which viewports have sync lock turned on?

File: packages/studio/index.d.ts (hunk 1 above)
Comment (defunctzombie) [reply packages/studio/index.d.ts]: Marking as resolved as this is no longer part of the public extension interface.

File: packages/studio/index.d.ts [hunk 3]
```diff
@@ -261,6 +267,11 @@ declare module "@foxglove/studio" {
      */
     setParameter: (name: string, value: ParameterValue) => void;
 
+    /**
+     * Set the transient state shared by panels of the same type as the caller of this function.
+     */
```
Comment (defunctzombie) [inline packages/studio/index.d.ts:272]: Is it worth noting that this will _not_ be saved?

File: packages/studio-base/src/providers/PanelStateContextProvider.tsx (hunk 2 above)
Comment (foxymiles) [reply packages/studio-base/src/providers/PanelStateContextProvider.tsx:111]: Good question. I think I can add this without too much extra complexity.

**Comment by defunctzombie** [comment]
These APIs make sense to me.

Once we are in a place where we are happy with the APIs I think it would be good to open a website PR to add the docs so we get a sense of how it reads.

**Comment by jtbandes** [comment]
> `    setSharedPanelState: (state: undefined | Readonly<string, unknown>) => void;`

Should we offer a functional form of `set((old) => new)`? Or maybe this doesn't provide much real world benefit?

**Comment by foxymiles** [comment]
> Should we offer a functional form of `set((old) => new)`? Or maybe this doesn't provide much real world benefit?

Generally I like having that as an option although it does seem a bit weird to use it for this particular case. It's easy enough to add.

File: packages/studio/index.d.ts (hunk 3 above)
Comment (foxymiles) [reply packages/studio/index.d.ts:273]: Calleing this out specifically in the comment now.

File: packages/studio-base/src/providers/PanelStateContextProvider.tsx (hunk 2 above)
Comment (foxymiles) [reply packages/studio-base/src/providers/PanelStateContextProvider.tsx:131]: I've added some garbage collection logic around shared state to clear it out once there is no longer a panel of the corresponding type using it.

File: packages/studio-base/src/panels/ThreeDeeRender/ThreeDeeRender.tsx [hunk 4]
```diff
@@ -819,13 +886,16 @@ export function ThreeDeeRender({ context }: { context: PanelExtensionContext }):
...
 
   const onTogglePerspective = useCallback(() => {
-    setConfig((prevConfig) => ({
-      ...prevConfig,
-      cameraState: { ...prevConfig.cameraState, perspective: !prevConfig.cameraState.perspective },
-    }));
-    // Wait for the setConfig to propagate to the renderer before updating the settings tree
-    setTimeout(() => renderer?.updateCoreSettings(), 0);
-  }, [renderer]);
+    const currentState = renderer?.getCameraState().perspective ?? false;
```
Comment (snosenzo) [inline packages/studio-base/src/panels/ThreeDeeRender/ThreeDeeRender.tsx:889]: This does look like a less hacky solution without the timeout, but why this change?
just unclear on when to use `actionHandler` vs `setConfig` and when to use which

**Review by snosenzo — APPROVED** [review]
looks good to me, but might want more seasoned eyes on it for some of the `renderState` stuff since I'm a little less familiar

File: packages/studio-base/src/components/PanelExtensionAdapter/renderState.ts [hunk 5]
```diff
@@ -144,6 +147,14 @@ function initRenderStateBuilder(): BuildRenderStateFn {
       }
     }
 
+    if (watchedFields.has("sharedPanelState")) {
+      if (sharedPanelState !== prevSharedPanelState) {
+        shouldRender = true;
+        prevSharedPanelState = sharedPanelState;
+        renderState.sharedPanelState = { ...sharedPanelState };
```
Comment (jtbandes) [inline packages/studio-base/src/components/PanelExtensionAdapter/renderState.ts:154]: why shallow copy?

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/CoreSettings.ts [hunk 6]
```diff
@@ -150,6 +150,12 @@ export class CoreSettings extends SceneExtension {
                   ? "This setting requires a restart to take effect"
                   : undefined,
             },
+            syncCamera: {
+              label: "Sync camera",
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/CoreSettings.ts:154]: Can we add help text that briefly describes how the setting works?

File: packages/studio-base/src/providers/PanelStateContextProvider.tsx [hunk 7]
```diff
@@ -61,8 +75,101 @@ export function usePanelSettingsTreeUpdate(): (newTree: ImmutableSettingsTree) =
...
+const selectCurrentLayoutId = (state: LayoutState) => state.selectedLayout?.id;
+
+const selectLayoutConfigById = createHasSameKeySelector(
+  (state: LayoutState) => state.selectedLayout?.data?.configById,
+  (config) => config,
+);
+
+const selectPanelTypesInUse = createSelector(selectLayoutConfigById, (config) => {
+  return uniq(Object.keys(config ?? {}).map(getPanelTypeFromId));
+});
```
Comment (jtbandes) [inline packages/studio-base/src/providers/PanelStateContextProvider.tsx:156]: Isn't it possible that there are panel types in use that don't have any saved config, but are using the shared state API?

**Review by jtbandes — APPROVED** [review]

File: packages/studio-base/src/providers/PanelStateContextProvider.tsx (hunk 7 above)
Comment (foxymiles) [reply packages/studio-base/src/providers/PanelStateContextProvider.tsx:156]: It looks like we still create a (empty) key in `configById` for panels with no settings.

File: packages/studio-base/src/components/PanelExtensionAdapter/renderState.ts (hunk 5 above)
Comment (foxymiles) [reply packages/studio-base/src/components/PanelExtensionAdapter/renderState.ts:154]: You mean instead of just passing it through? I guess you're right it's not really necessary. I'll take it out.
